### ISSUE TITLE
establish_connection breaking foreign_key for nested model associations

### ISSUE THREAD

**Issue by viniciusfont** [rails/rails#25091]
### Steps to reproduce

rails new Test -d postgresql
rails g Model Restaurant
rails g Model Section restaurant:references
add establish_connection to the 2 models to use a diferent database
### Expected behavior

I expected that I could create a Restaurant with a nested form
### Actual behavior

PG::ForeignKeyViolation: ERROR: insert or update on table "sections" violates foreign key constraint "fk_rails_14e0e2a999" DETAIL: Key (restaurant_id)=(4) is not present in table "restaurants". : INSERT INTO "sections" ("title_input_id", "restaurant_id", "created_at", "updated_at") VALUES ($1, $2, $3, $4) RETURNING "id"
### System configuration

**Rails version**: 4.2.6

**Ruby version**: 2.2.0
### More Info

```
class Restaurant < ActiveRecord::Base
  establish_connection :'f7-api'

  has_many :sections, dependent: :destroy
  has_many :items, through: :sections
  accepts_nested_attributes_for :sections, allow_destroy: true
end

class Section < ActiveRecord::Base
  establish_connection :'f7-api'

  belongs_to :restaurant
  has_many :items, dependent: :destroy
  has_many :options, through: :items
  accepts_nested_attributes_for :items, allow_destroy: true
end

class RestaurantsController < ApplicationController
# POST /restaurants
  # POST /restaurants.json
  def create
    @restaurant = Restaurant.new(restaurant_params)

    respond_to do |format|
      if @restaurant.save
        format.html { redirect_to @restaurant, notice: 'Restaurant was successfully created.' }
        format.json { render :show, status: :created, location: @restaurant }
      else
        format.html { render :new }
        format.json { render json: @restaurant.errors, status: :unprocessable_entity }
      end
    end
  end

def restaurant_params
      params.require(:restaurant).permit(
        :id,
        :name,
        :business_name,
        :cnpj,
        :description,
        :phone,
        :mobile,
        :website,
        :user_id,
        :street,
        :complement,
        :number,
        :lat,
        :lng,
        :zip_code,
        :neighborhood_id,
        :city_id,
        :state_id,
        :country_id,
        photos: [:id, :image, :main],
        open_hours: [:id, :weekday, :opens_at, :closes_at],
        cuisine_ids: [],
        category_ids: [],
        sections_attributes: [
          :id,
          :title,
          :restaurant_id,
          :_destroy,
          items_attributes: [
            :id,
            :title,
            {:description => []},
            :section_id,
            :price_cents,
            :not_equal,
            :_destroy,
            options_attributes: [
              :id,
              {:description => []},
              :item_id,
              :price_cents,
              :_destroy
            ]
          ]
        ]
      )
    end
end
```

**Comment by vipulnsward** [comment]
What is order of insertion you see?
Since you are probably inserting in a different database and expecting it to persist in some other database at the same time, it would raise an error.
Can you confirm what these 2 different databases are, master/slave/read-only etc.

**Comment by viniciusfont** [comment]
I'ts a database from another rails application with the same postgres user.

**Comment by viniciusfont** [comment]
The solution is to add `foreign_key: 'id'` to the `has_many` in the `Restaurant` model.

I dont know why without this doesnt work.

`has_many :sections, dependent: :destroy, foreign_key: 'id'`

**Comment by viniciusfont** [comment]
Why using other database connection break the postgres foreign_key relation?

**Comment by sbonami** [comment]
I can confirm that I'm seeing this in Rails 4.2, and is reproducible from the console as well.

**Comment by sbonami** [comment]
I was able to achieve the result I was looking for by overriding the `DEFAULT_ENV` constant rather than explicitly establishing connections with each model. I'm not certain of the implications of doing such as of yet, but it did achieve the desired result while allowing nested association persistence to operate as expected.

**Comment by paramadeep** [comment]
I am running rails 5.1.4. Getting the following exception while trying to create ans save a nested model object (models use establish connection)
```
ActiveRecord::InvalidForeignKey (PG::ForeignKeyViolation: ERROR:  insert or update on table "food_groups" violates foreign key constraint "food_groups_food_id_fkey"
DETAIL:  Key (food_id)=(11393) is not present in table "food_items".
```

**Comment by sbonami** [comment]
I ended up tracking this down and solving it in a more "rails" way. So, without overriding a constant (which is shameful and appalling), set your parent model (or a higher-level base model) as an [abstract model](http://www.informit.com/articles/article.aspx?p=2220311&seqNum=5) and call `establish_connection` there. This creates a shared connection pool, rather than a newly established connection pool per model.

I surmised that the issue comes from 2 different transactions on 2 different connections, hence an inability to see the related model.
